apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/plugin/WitnessClassFinder.java

```diff
@@ -36,6 +36,17 @@
      * @return true, if the given witnessClass exists, through the given classLoader.
      */
     public boolean exist(String witnessClass, ClassLoader classLoader) {
+        return getResolution(witnessClass, classLoader)
```

Comment:
@muse-dev The `hashmap` modification is already in the double-check and `synchronized`. I don't think this is an issue.

Reply:
@TomMD Could you double-check about this?

Reply:
> Do I need to care about code review of muse-dev?

@libinglong I am not sure whether this is an issue. Pinged the muse-dev team. But at least, this is an issue brought by this PR.

Reply:
@wu-sheng I haven't looked at the actual trace produced by the tool but here are my thoughts - please let me know if they seem off and I can ask the tool for its explanation.

The message here is due to the same core issue as you see in the `getResolution` comment below.  When calling `getResolution` then without synchronization we execute:

```
 TypePool typePool = poolMap.get(mappingKey);
```

And that can race with a write to `poolMap` occurring in another thread.  There might not be a `mappingKey` inside of `poolMap` because `poolMap` could be modified prior to calling `get`.  If you'd like to be sure the element exists then you'd need the synchronize block to encompass both the `get` and the `containsKey` call.

We're planning on making traces accessible from the comments with a commands to the bot - this would illuminate the supposed bugs and help developers decide if the trace is feasible and worth changing.  We welcome any thoughts you have about bug explanation and elaboration.

Reply:
@TomMD Typical race condition of HashMap should like this, https://stackoverflow.com/questions/7830791/java-hashmap-race-condition.

I can't find the `get` conflicts with `put`. What do I miss?

Reply:
Maybe It should not be assumed that the put method for a key has no effect on the get method for another key In concurrent environment as long as the data struct(hashmap) is thread unsafe.

Reply:
@wu-sheng I don't believe you miss anything, it's merely a difference of what one considers a race.  If you `get` from a variable outside of synchronized, and that variable is sometimes used in a `synchronized` block then you receive this read/write race warning.

Your read might race with a write because the lock isn't held.  You are probably unconcerned and can quickly dismiss the warning in this case.  The link you posted suggests you are concern is about races regarding the handling of the hashmap itself (i.e. the collision and resize issues in the answer) and it suffices to avoid write/write races.

I believe in practice read/write races are concerning almost exclusively in multi-variable situations where there are invariants between the variables that are broken within the writers so the readers must ensure they see a consistent state, usually with a lock.

Reply:
@TomMD My understanding is, Java's HashMap is only facing write/write race, but here, it shows me the `HashMap#get` will need a lock. Is this correct?